### DISCUSSION TITLE
Disable eager loading for some includes

### DISCUSSION THREAD

**Discussion by cnelson-ma** [spatie/laravel-query-builder#500] [Q&A] [answered]
At the moment the query builder will try to eager load all parameters passed via the `?include=` query parameter. The eager loading occurs weather you specify `allowedIncludes` or not.

This tries to eager load all includes via the `PermissionGroup` model, as I would expect. I receive an error because `PermissionGroup` has no `permission_tree `method:

```php
// api/permission-group?include=permission_tree,permissions,users
...
QueryBuilder::for(PermissionGroup::class)
            ->allowedIncludes(['permission_tree', 'permissions', 'users'])
            ->findOrFail($id);
...
```

I remove the allowed includes now to stop the eager loading and will handle the validation my self. I expect that this will not try and eager load the include string.

However it still tries to eager load all includes:

```php
// api/permission-group?include=permission_tree,permissions,users
...
QueryBuilder::for(PermissionGroup::class)
            ->findOrFail($id);
...
```

There is often a need to allow for includes that are arbitrary data that is defined in the transformer that is not apart of a relation.

Fractal/Dingo is what we are using, which seems to be a common combination. 

Dingo's adapter to Fractal allows you to specify lazy loading includes here on the transformation adapter: https://github.com/dingo/api/blob/master/src/Transformer/Adapter/Fractal.php#L207

Which was merged because of this discussion: https://github.com/dingo/api/pull/1036

Is there a way to disable this that I am missing? This is make or break for us.

**Comment by cnelson-ma** [comment]
It looks like there is even a TODO for this: https://github.com/spatie/laravel-query-builder/blob/master/src/Concerns/AddsIncludesToQuery.php#L81

**Comment by dominikb** [comment, marked as answer]
Hi @cnelson-ma,

thanks for opening up a discussion.

> The eager loading occurs weather you specify `allowedIncludes` or not.

This should not happen. There are two different cases:
1. No `allowedIncludes()` call is made at all -> every include parameter gets ignored silently
2. `allowedIncludes()` does not contain a user-supplied parameter -> an `InvalidIncludeQuery` gets thrown.

Can you try to isolate the issue in a failing test case? I'd be happy to fix this issue, once isolated.

>It looks like there is even a TODO for this: https://github.com/spatie/laravel-query-builder/blob/master/src/Concerns/AddsIncludesToQuery.php#L81

This TODO likely exists to allow us to give the user a more meaningful exception if an unknown relation is included. Currently, we let the underlying QueryBuilder check for this, where an `Illuminate\Database\Eloquent\RelationNotFoundException` gets thrown. I may be wrong on this one though.

**Reply by cnelson-ma** [reply]
Sorry for taking your time on this. There were some interactions happening in Dingo/Fractal that I did not fully understand. 

You are correct. Your system is working as you stated!

Thanks for the prompt help on this. I have been banging my head getting all of this setup and this package is amazing for us.
